oap-server/server-telemetry/telemetry-prometheus/src/main/java/org/apache/skywalking/oap/server/telemetry/prometheus/PrometheusMetricsCollector.java

```diff
@@ -23,23 +23,91 @@
-    @Override public Iterable<MetricFamily> collect() {
+    @Override
+    public Iterable<MetricFamily> collect() {
         Enumeration<Collector.MetricFamilySamples> mfs = CollectorRegistry.defaultRegistry.metricFamilySamples();
         List<MetricFamily> result = new ArrayList<>();
         while (mfs.hasMoreElements()) {
             Collector.MetricFamilySamples metricFamilySamples = mfs.nextElement();
             List<MetricFamily.Sample> samples = new ArrayList<>(metricFamilySamples.samples.size());
-            MetricFamily m = new MetricFamily(metricFamilySamples.name, MetricFamily.Type.valueOf(metricFamilySamples.type
-                .name()), metricFamilySamples.help, samples);
+            MetricFamily m = new MetricFamily(
+                metricFamilySamples.name, MetricFamily.Type.valueOf(metricFamilySamples.type
+                                                                        .name()), metricFamilySamples.help, samples
+            );
             result.add(m);
             for (Collector.MetricFamilySamples.Sample sample : metricFamilySamples.samples) {
-                samples.add(new MetricFamily.Sample(sample.name, sample.labelNames, sample.labelValues, sample.value, sample.timestampMs));
+                samples.add(new MetricFamily.Sample(
+                    sample.name, sample.labelNames, sample.labelValues, sample.value,
+                    sample.timestampMs
+                ));
             }
         }
         return result;
     }
+
+    @Override
+    public long heapMemoryUsage() {
+        final AtomicLong heapMemoryUsed = new AtomicLong();
+        this.find("jvm_memory_bytes_used")
+            .ifPresent(metricFamily -> {
+                metricFamily.samples.forEach(sample -> {
+                    for (int i = 0; i < sample.labelNames.size(); i++) {
+                        if (sample.labelNames.get(i).equals("area") && sample.labelValues.get(i)
+                                                                                         .equals("heap")) {
+                            heapMemoryUsed.addAndGet((long) sample.value);
+                        }
+                    }
+                });
+            });
+        return heapMemoryUsed.longValue();
+    }
+
+    @Override
+    public long heapMemoryMax() {
+        final AtomicLong heapMemoryMax = new AtomicLong();
+        this.find("jvm_memory_bytes_max")
+            .ifPresent(metricFamily -> {
+                metricFamily.samples.forEach(sample -> {
+                    for (int i = 0; i < sample.labelNames.size(); i++) {
+                        if (sample.labelNames.get(i).equals("area") && sample.labelValues.get(i)
+                                                                                         .equals("heap")) {
+                            heapMemoryMax.addAndGet((long) sample.value);
+                        }
+                    }
+                });
+            });
```

Comment:
And this looks poorly performant, it is complexity of `#metrics family * #labels`, cases might be the metrics family or the labels are huge and this is excuted every 10 seconds, which might be an issue itself. Do we have a more performant way to find the metrics family or labels ?

Reply:
This would not goes into all metrics x all labels, it does metric name match first, then with labels.
AFAIK the number of labels of JVM memory metric should be limited.

We don't need to worry about its performance?